File: 4h-unfolder-win/src/FourHUnfolder.Geometry/Algorithms/BoundaryPolygonComputer.cs
using System.Numerics;
using FourHUnfolder.Domain.Results;

namespace FourHUnfolder.Geometry.Algorithms;

/// <summary>
/// Computes the outer boundary polygon of a group of unfolded faces by collecting
/// all non-fold edges and chaining them into an ordered closed polygon path.
/// </summary>
public static class BoundaryPolygonComputer
{
    private const float SnapEps = 0.001f;

    /// Returns the ordered boundary polygon in the same coordinate space as the faces,
    /// or null for degenerate/empty input.
    public static List<Vector2>? Compute(IReadOnlyList<UnfoldedFace> faces)
    {
        if (faces.Count == 0) return null;

        var edges     = new List<(Vector2 A, Vector2 B)>();
        var seenById  = new HashSet<int>();
        var seenByCoord = new HashSet<(long, long, long, long)>();

        foreach (var face in faces)
        {
            var v = face.Vertices;
            for (int i = 0; i < 3; i++)
            {
                if (face.EdgeIsFold[i]) continue;

                int meshId = face.MeshEdgeIds[i];
                if (meshId >= 0)
                {
                    if (!seenById.Add(meshId)) continue;
                }
                else
                {
                    var key = CoordKey(v[i], v[(i + 1) % 3]);
                    if (!seenByCoord.Add(key)) continue;
                }
                edges.Add((v[i], v[(i + 1) % 3]));
            }
        }

        return ChainEdges(edges);
    }

    private static List<Vector2>? ChainEdges(List<(Vector2 A, Vector2 B)> edges)
    {
        if (edges.Count == 0) return null;

        var polygon   = new List<Vector2> { edges[0].A, edges[0].B };
        var remaining = new List<(Vector2 A, Vector2 B)>(edges);
        remaining.RemoveAt(0);

        for (int guard = 0; guard < edges.Count && remaining.Count > 0; guard++)
        {
            var tail  = polygon[^1];
            bool found = false;
            for (int k = 0; k < remaining.Count; k++)
            {
                var (a, b) = remaining[k];
                if (Near(tail, a)) { polygon.Add(b); remaining.RemoveAt(k); found = true; break; }
                if (Near(tail, b)) { polygon.Add(a); remaining.RemoveAt(k); found = true; break; }
            }
            if (!found) break;
        }

        return polygon.Count >= 3 ? polygon : null;
    }

    private static bool Near(Vector2 a, Vector2 b)
    {
        var d = a - b;
        return d.X * d.X + d.Y * d.Y < SnapEps * SnapEps;
    }

    // Symmetric (direction-independent) rounded coordinate key for deduplication.
    private static (long, long, long, long) CoordKey(Vector2 a, Vector2 b)
    {
        long ax = (long)MathF.Round(a.X * 1000), ay = (long)MathF.Round(a.Y * 1000);
        long bx = (long)MathF.Round(b.X * 1000), by = (long)MathF.Round(b.Y * 1000);
        return ax < bx || (ax == bx && ay <= by) ? (ax, ay, bx, by) : (bx, by, ax, ay);
    }
}

File: 4h-unfolder-mac-swift/Sources/FourHUnfolderCore/Core/Algorithms/BoundaryPolygonComputer.swift
import simd

// MARK: - Piece outer-boundary polygon (backlog Phase 1: wire Outline Padding into export/canvas)
//
// Port of Windows FourHUnfolder.Geometry.Algorithms.BoundaryPolygonComputer.Compute/ChainEdges.
// Collects a piece's non-fold edges (cut + mesh-boundary — both get physically cut, so both are
// part of the piece's outer silhouette) and chains them into an ordered closed loop. This is the
// missing piece that left `PolygonOffset.inflate` with nothing to consume: that function inflates
// one already-ordered closed polygon, but macOS had no piece-boundary tracer to hand it one.

enum BoundaryPolygonComputer {

    private static let snapEpsMm: Float = 0.001

    /// Returns the ordered boundary polygon of a set of faces (one piece), or nil for
    /// degenerate/empty input. `vertsFor` supplies each face's triangle vertices — pass the
    /// default (raw `face.v0/v1/v2`) for export, where `result.faces` already holds final
    /// baked positions, or a transform closure (e.g. canvas's `effectiveVerts`) to trace the
    /// boundary in a piece's live rotated/offset position.
    static func compute(
        faces: [UnfoldedFace],
        vertsFor: (UnfoldedFace) -> (SIMD2<Float>, SIMD2<Float>, SIMD2<Float>) = { ($0.v0, $0.v1, $0.v2) }
    ) -> [SIMD2<Float>]? {
        guard !faces.isEmpty else { return nil }

        var edges: [(SIMD2<Float>, SIMD2<Float>)] = []
        var seenById = Set<Int>()
        var seenByCoord = Set<CoordKey>()

        for face in faces {
            let v = vertsFor(face)
            let verts = [v.0, v.1, v.2]
            for ei in 0..<3 where !face.edgeIsFold(ei) {
                let meshId = face.meshEdgeId(ei)
                if meshId >= 0 {
                    guard seenById.insert(meshId).inserted else { continue }
                } else {
                    let key = coordKey(verts[ei], verts[(ei + 1) % 3])
                    guard seenByCoord.insert(key).inserted else { continue }
                }
                edges.append((verts[ei], verts[(ei + 1) % 3]))
            }
        }

        return chainEdges(edges)
    }

    // MARK: - Helpers

    private static func chainEdges(_ edges: [(SIMD2<Float>, SIMD2<Float>)]) -> [SIMD2<Float>]? {
        guard !edges.isEmpty else { return nil }

        var polygon: [SIMD2<Float>] = [edges[0].0, edges[0].1]
        var remaining = edges
        remaining.removeFirst()

        var guardCount = 0
        while guardCount < edges.count && !remaining.isEmpty {
            guardCount += 1
            let tail = polygon[polygon.count - 1]
            guard let k = remaining.firstIndex(where: { near($0.0, tail) || near($0.1, tail) }) else { break }
            let (a, b) = remaining[k]
            polygon.append(near(a, tail) ? b : a)
            remaining.remove(at: k)
        }

        // When the chain closes the loop (last point lands back on the first), C#'s identical
        // ChainEdges leaves that duplicate in place — harmless there because Clipper2 silently
        // strips duplicate/collinear points before offsetting. `PolygonOffset.inflate` has no such
        // cleanup (documented as local-joins-only, Clipper-free), so a leftover duplicate makes it
        // treat that shared corner as two degenerate zero-length edges and drop the corner's arc
        // entirely, leaving a flat notch. Trim it here instead so `inflate` sees one clean corner.
        if polygon.count >= 4, near(polygon[0], polygon[polygon.count - 1]) {
            polygon.removeLast()
        }

        return polygon.count >= 3 ? polygon : nil
    }

    private static func near(_ a: SIMD2<Float>, _ b: SIMD2<Float>) -> Bool {
        simd_length_squared(a - b) < snapEpsMm * snapEpsMm
    }

    /// Direction-independent rounded coordinate key for edge dedup when no meshEdgeId is known.
    private struct CoordKey: Hashable {
        let ax, ay, bx, by: Int64
    }

    private static func coordKey(_ a: SIMD2<Float>, _ b: SIMD2<Float>) -> CoordKey {
        let ax = Int64((a.x * 1000).rounded()), ay = Int64((a.y * 1000).rounded())
        let bx = Int64((b.x * 1000).rounded()), by = Int64((b.y * 1000).rounded())
        return ax < bx || (ax == bx && ay <= by)
            ? CoordKey(ax: ax, ay: ay, bx: bx, by: by)
            : CoordKey(ax: bx, ay: by, bx: ax, by: ay)
    }
}
